### PR TITLE
Implement Hash#path_to_key

### PR DESCRIPTION
**Pull request by rudy-on-rails**
This PR is a first attempt to contribute to MRI, done at the RubyHackChallenge at Cookpad office in Bristol. I apologize in advance for anything that might be missing here, in relation to how this should be approached. 

It has been started with https://github.com/ko1/rubyhackchallenge/issues/41. After the implementation, me and @skateman had a chart with @matz about how applicable would it be to have that code merged.

Matz suggested that a method like this, described afterwards in https://github.com/ko1/rubyhackchallenge/issues/44, would have more applicability and just generally be more useful.

### CODE REVIEW

**Comment by rudywseidinger** [comment]
@skateman This is the implementation I came up with.

**Comment by nobu** [comment]
Where `h = {a: 1}`, both of `h.path_to_key(:a)` and `h.path_to_key(:b)` return `[]`.
Is this intentional?

For recursive hash, `h = {}; h[:x] = h`, `h.path_to_key(:a)` returns `[:x]`, but `h` doesn't contain `:a`.

**Comment by rudywseidinger** [comment]
@nobu Interesting comment....
Number 1 is intentional, yes, although might not be the most ideal way of exposing that the path is unexistent _or_ it's in the first-level (root) of the Hash object.

About number two, I will add a test for this case and fix it.

File: hash.c [hunk 1]
```diff
@@ -4510,6 +4510,52 @@ env_update(VALUE env, VALUE hash)
...
+
+    VALUE keys;
+    VALUE curr_key;
+    VALUE curr_element;
+
+    while(!rb_eql(hash, Qnil) && !rb_hash_eql(prev_hsh, hash)){
+        keys = rb_hash_keys(hash);
+        prev_hsh = hash;
+
+        for (int i = 0; i < RARRAY_LEN(keys); ++i)
```
Comment (nobu) [inline hash.c]: declaration in `for` is not C89.

**Comment by nobu** [comment]
Why not `nil` when the key does not exist?

**Comment by rudy-on-rails** [comment]
I think it might make more sense, @nobu . I will check it out on how can I approach that and change it.

**Comment by nobu** [comment]
I guess that you'd expect `{a: {b: 0, x: 1}}.path_to_key(:x)` to be `[:a]`.

**Comment by rudy-on-rails** [comment]
Yes @nobu. Just fixed that and another recursive hash issue...

**Comment by rudy-on-rails** [comment]
I will have to slightly change the implementation of this, because hashes like

```ruby
hsh = {:x=>{:e=>1}, :y=>[1, 2, 3], :z=>{:a=>{:w=>100}}}
hsh.path_to_key(:w)
```
will be expected to return `[:z, :w]` but are not in the moment due to the way sub-hashes are analised.

**Comment by nobu** [comment]
`{a: {}, b: {x: 1}}.path_to_key(:x)` returns `nil`.

**Comment by nobu** [comment]
`h = {a: {}}; h[:a][:b] = h; h.path_to_key(:x)` doesn't return.

**Comment by nobu** [comment]
What do you expect when a hash has multiple paths to the same key, e.g., `{a: {x: 0}, b: {x: 1}}.path_to_key(:x)`?

**Comment by rudy-on-rails** [comment]
The right expectation would be [:a], the first path found that matches.

**Comment by rudy-on-rails** [comment]
Hey @nobu. Can you re-check this? I have re-implemented using a recursive approach, but I don't know how much my recursion design might hurt some other parts. I've used some of the recursions over the hash as base. Cheers

**Comment by hsbt** [comment]
@rudy-on-rails Can you address this feature request to https://bugs.ruby-lang.org?

We do not handle new feature on GitHub.

**Comment by rudy-on-rails** [comment]
Hey @hsbt , sure thing. I will write the rationale there.

**Comment by rudy-on-rails** [comment]
Done @hsbt https://bugs.ruby-lang.org/issues/14636

**Comment by k0kubun** [comment]
It seems to have a conflict now. Could you rebase this from master?

**Comment by k0kubun** [comment]
Let me close this as it has not been updated for a while. Please reopen this after resolving conflicts. Thanks.

**Pull request closed without merging**